src/frontend/Preprocessor.ml

```diff
@@ -5,16 +5,39 @@ open Lexing
 open Debugging
 module Str = Re.Str
 
-let dup_exists l =
-  match List.find_a_dup ~compare:String.compare l with
-  | Some _ -> true
-  | None -> false
-
 let include_stack = Stack.create ()
 let include_paths : string list ref = ref []
 let included_files : string list ref = ref []
 let size () = Stack.length include_stack
 
+let locations_map : (string * Middle.Location.t option) String.Table.t =
+  String.Table.create ()
+
+let reset_locations () = Hashtbl.clear locations_map
+
+let new_file_start_position filename included_from =
+  if Option.is_none included_from then
+    {Lexing.pos_fname= filename; pos_lnum= 1; pos_bol= 0; pos_cnum= 0}
+  else
+    let key = "\u{0}" ^ string_of_int (Hashtbl.length locations_map) in
```

Comment:
Why do we need the key to start with NULL?

Reply:
To distinguish it from  a filename (which never begins with NUL, I hope)

There's a micro-optimization here that I'm not sure is actually worth it. (I didn't profile the code or run benchmarks)
My reasoning is that most Stan code lives in the toplevel file, not the `#include`d ones. In that case there's no `included_from` and it would be better to just store the filename directly in `pos_fname` instead of going through the hash table. So that's what we do! Then we just just need a fast way to recognize which `pos_fname` really are hash table keys and which are plain filenames. The easiest way to do that is to tag all keys with a character that is unlikely to occur in a filename. 

I can add a comment explaining this--or just remove the "toplevel fast path"; while a NUL check is surely faster than a hash table lookup I don't know if the difference is noticeable.

Reply:
I did some minimal profiling of this branch, and then again with this additional logic commented out. 

It does seem to be faster for a model with no `#include`s (168.22K vs 281.29K total cycles spent in the `location_of_position` function) to have this specialized behavior, but not in a huge way (the existing `Location.of_position_opt` function is 175.6K cycles for this model, and the total runtime of the compiler is 93 million cycles)

So far as I can tell, no platform allows a filename to contain NUL, so it's probably fine, but I think a comment would go a long way.

Reply:
Huh, for some reason `Preprocessor.location_span_of_position` is showing up as expensive in the profiling I'm doing, even though all it does is call `location_of_position` twice and return. 

Odd, but the overall time on `Parse.parse` is essentially the same for a model with no `#include`s and much (up to 10x) faster for a model which is just one big `#include`. I imagine it would be even better if you have `includes` which themselves have further `includes`.